### PR TITLE
feat: change avoid-returning-widgets

### PR DESCRIPTION
**Pull request by incendial**
### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix
[ ] New rule
[X] Changes an existing rule
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

### What changes did you make? (Give an overview)
As mentioned here https://github.com/dart-lang/linter/pull/2582#issuecomment-819723672 we should not lint every class method with `Widget` return type, because there are cases where it's totally valid to have a `Widget`-returning method and it's used in builders. So this request supports this approach for methods, but still lint getters and global functions as before.

### CODE REVIEW

**Comment by incendial** [comment]
@roman-petrov could you please a look too? I'd very happy to get you feedback about this.

File: lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_returning_widgets/visit_declaration.dart [hunk 1]
```diff
@@ -0,0 +1,72 @@
+part of 'avoid_returning_widgets.dart';
+
+Declaration? _visitDeclaration(
+  Declaration node,
+  SimpleIdentifier name,
+  TypeAnnotation? returnType,
+  Iterable<String> _ignoredNames,
```
Comment (dkrutskikh) [inline lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_returning_widgets/visit_declaration.dart:7]: @incendial argument name started with underscore looks strange

File: lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_returning_widgets/visit_declaration.dart [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  Iterable<String> _ignoredAnnotations, {
+  required bool isSetter,
+}) {
+  final hasIgnoredAnnotation = node.metadata.firstWhereOrNull(
```
Comment (dkrutskikh) [inline lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_returning_widgets/visit_declaration.dart:11]: @incendial you can use `any` instead `firstWhereOrNull`

File: lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_returning_widgets/visit_declaration.dart [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  Iterable<String> _ignoredNames,
+) =>
+    name == 'build' || _ignoredNames.contains(name);
+
+bool _isWidget(DartType? type) =>
+    type?.getDisplayString(withNullability: false) == 'Widget';
+
+bool _isSubclassOfWidget(DartType? type) =>
+    type is InterfaceType &&
+    type.allSupertypes.firstWhereOrNull(_isWidget) != null;
```
Comment (dkrutskikh) [inline lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_returning_widgets/visit_declaration.dart:66]: @incendial any instead firstWhereOrNull

**Review by dkrutskikh — CHANGES_REQUESTED** [review]

**Comment by dkrutskikh** [comment]
@incendial please add test cases for uncovered cases
https://codecov.io/gh/dart-code-checker/dart-code-metrics/src/improve-avoid-returning-widgets/lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_returning_widgets/visitor.dart

https://codecov.io/gh/dart-code-checker/dart-code-metrics/src/improve-avoid-returning-widgets/lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_returning_widgets/visit_declaration.dart

File: lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_returning_widgets/visit_declaration.dart (hunk 1 above)
Comment (incendial) [reply lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_returning_widgets/visit_declaration.dart]: Fixed.

File: lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_returning_widgets/visit_declaration.dart (hunk 2 above)
Comment (incendial) [reply lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_returning_widgets/visit_declaration.dart]: Fixed.

File: lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_returning_widgets/visit_declaration.dart (hunk 3 above)
Comment (incendial) [reply lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_returning_widgets/visit_declaration.dart]: Fixed.

**Review by dkrutskikh — APPROVED** [review]

**Comment by roman-petrov** [comment]
@incendial sorry for late response, you mentioned me above but I missed @mention notification in my mail. I will try to improve my email filters to solve this, I am still quite interesting in contributing to this project. 

This specific update logically looks quite good to me, returning widgets from getters for me looks even worse than returning from functions.

**Comment by incendial** [comment]
> @incendial sorry for late response, you mentioned me above but I missed @mention notification in my mail.

No problem.

> This specific update logically looks quite good to me, returning widgets from getters for me looks even worse than returning from functions.

Thanks for taking a look!

**Pull request merged**